### DISCUSSION TITLE
Get address of external function

### DISCUSSION THREAD

**Discussion by mfthomps** [NationalSecurityAgency/ghidra#6080] [Q&A] [answered]
My plugin creates a map of function names to addresses for use by an external simulator when resolving a call instruction to its function name.  However, when attempting to get addresses for external functions, i.e., when iterating over function_manager.getExternalFunctions(), the addresses returned by function.getBody() are null.  How can I get the  address of each external function, i.e., the value in the call instruction?

**Comment by dev747368** [comment]
```
>>> currentProgram.functionManager.getExternalFunctions().next()
<EXTERNAL>::printf
>>> currentProgram.functionManager.getExternalFunctions().next().entryPoint
EXTERNAL:00000001
>>> currentProgram.functionManager.getExternalFunctions().next().body
[empty]
```
External functions aren't real functions that have a body (as seen by the python snippet above), but they do have an address (entry point) in a fake address space ("EXTERNAL").

The details of how that function work will depend on the binary, could be a thunk, or an indirect pointer, etc.

**Comment by mfthomps** [comment]
I had also tried function.getExternalLocation().getAddress() with no luck.  It is not a thunk.
In the example I'm looking at, it is a call to a windows dll pointer.   
To clarify, I'm not looking for the address in the target library.  I'm looking for the address that appears in the call instruction that then eventually leads to the external function.   I want to translate "call 0xdeadbeef" to "call printf" or whatever.

**Comment by mfthomps** [comment]
I had tried variations of that in Java:
```
                Symbol s = f.getSymbol();
                Address addr = s.getAddress();
```
But the address offset is always an index value (incrementing integers).  Also tried with getExternalSymbol on the SymbolTable.
If I can get the address of the start of the list of pointers to external functions, I expect I can use that index to compute an offset from that.
... oh, I see the previous response suggesting use of Symbols was removed.

**Reply by ryanmkurtz** [reply]
Yea, sorry, i realized it was wrong

**Comment by dev747368** [comment, marked as answer]
try
```
ghidra.app.nav.NavigationUtils.getExternalLinkageAddresses(currentProgram, f.getEntryPoint() )[0]
```
(where f is the external function, and its entry point is going to be something like EXTERNAL:00000NNN)

**Reply by mfthomps** [reply]
That works.  Thanks tons.  The support for this project is fantastic.
